src/graph/immutable_graph.cc

```diff
@@ -144,328 +97,118 @@ CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids, bool is_multigraph,
-          }
-        }
-      }
-      return false;
+      return aten::CSRHasDuplicate(adj_);
     });
 }
 
 CSR::EdgeArray CSR::OutEdges(dgl_id_t vid) const {
   CHECK(HasVertex(vid)) << "invalid vertex: " << vid;
-  const dgl_id_t* indptr_data = static_cast<dgl_id_t*>(indptr_->data);
-  const dgl_id_t* indices_data = static_cast<dgl_id_t*>(indices_->data);
-  const dgl_id_t* edge_ids_data = static_cast<dgl_id_t*>(edge_ids_->data);
-  const dgl_id_t off = indptr_data[vid];
-  const int64_t len = OutDegree(vid);
-  IdArray src = NewIdArray(len);
-  IdArray dst = NewIdArray(len);
-  IdArray eid = NewIdArray(len);
-  dgl_id_t* src_data = static_cast<dgl_id_t*>(src->data);
-  dgl_id_t* dst_data = static_cast<dgl_id_t*>(dst->data);
-  dgl_id_t* eid_data = static_cast<dgl_id_t*>(eid->data);
-  std::fill(src_data, src_data + len, vid);
-  std::copy(indices_data + off, indices_data + off + len, dst_data);
-  std::copy(edge_ids_data + off, edge_ids_data + off + len, eid_data);
-  return CSR::EdgeArray{src, dst, eid};
+  IdArray ret_dst = aten::CSRGetRowColumnIndices(adj_, vid);
+  IdArray ret_eid = aten::CSRGetRowData(adj_, vid);
+  IdArray ret_src = aten::Full(vid, ret_dst->shape[0], NumBits(), ret_dst->ctx);
+  return CSR::EdgeArray{ret_src, ret_dst, ret_eid};
 }
 
 CSR::EdgeArray CSR::OutEdges(IdArray vids) const {
   CHECK(IsValidIdArray(vids)) << "Invalid vertex id array.";
-  const dgl_id_t* indptr_data = static_cast<dgl_id_t*>(indptr_->data);
-  const dgl_id_t* indices_data = static_cast<dgl_id_t*>(indices_->data);
-  const dgl_id_t* edge_ids_data = static_cast<dgl_id_t*>(edge_ids_->data);
-  const auto len = vids->shape[0];
-  const dgl_id_t* vid_data = static_cast<dgl_id_t*>(vids->data);
-  int64_t rstlen = 0;
-  for (int64_t i = 0; i < len; ++i) {
-    dgl_id_t vid = vid_data[i];
-    CHECK(HasVertex(vid)) << "Invalid vertex: " << vid;
-    rstlen += OutDegree(vid);
-  }
-  IdArray src = NewIdArray(rstlen);
-  IdArray dst = NewIdArray(rstlen);
-  IdArray eid = NewIdArray(rstlen);
-  dgl_id_t* src_ptr = static_cast<dgl_id_t*>(src->data);
-  dgl_id_t* dst_ptr = static_cast<dgl_id_t*>(dst->data);
-  dgl_id_t* eid_ptr = static_cast<dgl_id_t*>(eid->data);
-  for (int64_t i = 0; i < len; ++i) {
-    const dgl_id_t vid = vid_data[i];
-    const dgl_id_t off = indptr_data[vid];
-    const int64_t deg = OutDegree(vid);
-    if (deg == 0)
-      continue;
-    const auto *succ = indices_data + off;
-    const auto *eids = edge_ids_data + off;
-    for (int64_t j = 0; j < deg; ++j) {
-      *(src_ptr++) = vid;
-      *(dst_ptr++) = succ[j];
-      *(eid_ptr++) = eids[j];
-    }
-  }
-  return CSR::EdgeArray{src, dst, eid};
+  auto csrsubmat = aten::CSRSliceRows(adj_, vids);
+  auto coosubmat = aten::CSRToCOO(csrsubmat, false);
+  // Note that the row id in the csr submat is relabled, so
+  // we need to recover it using an index select.
+  auto row = aten::IndexSelect(vids, coosubmat.row);
+  return CSR::EdgeArray{row, coosubmat.col, coosubmat.data};
```

Comment:
is this implementation much slower than before? now we have to reconstruct csr and then coo and remap the row ids to the original graph.

Reply:
No. They are the same. The new implementation just breakdown the original logic into three functions. The CSRToCOO (without shuffling) is very efficient. The column indices and data arrays are just shallow copy.